File: 904.cpp
#include <map>
#include <vector>
using namespace std;
class Solution {
public:
  int totalFruit(vector<int> &fruits) {
    int ans = 0, curr = 0;
    // storing the last index of a fruit-type
    map<int, int> lastIndex;
    for (int i = 0; i < fruits.size(); ++i) {
      lastIndex[fruits[i]] = i;
      // if there are more than 2 distinct fruit types.
      if (lastIndex.size() > 2) {
        int out = -1;
        // look in the last indices map for the fruit type that was last.
        // this will be the fruit type that isn't the current one and isn't the
        // previous one. Hence, fruits[i] and fruits[i - 1]
        for (auto &[type, _] : lastIndex) {
          if (type != fruits[i - 1] && type != fruits[i]) {
            out = type;
            break;
          }
        }
        curr = i - lastIndex[out];
        lastIndex.erase(out);
      } else {
        ++curr;
      }
      ans = max(ans, curr);
    }
    return ans;
  }
};

int main() {}

File: 904.py
class Solution:
    def totalFruit(self, fruits: list[int]) -> int:
        ans, curr = 0, 0
        lastIndex = {}
        for i in range(len(fruits)):
            lastIndex[fruits[i]] = i
            if len(lastIndex) > 2:
                out = -1
                for key in lastIndex:
                    if key != fruits[i] and key != fruits[i - 1]:
                        out = lastIndex[key]
                        break
                curr = i - out
                lastIndex.pop(fruits[out])
            else:
                curr += 1
            ans = max(ans, curr)
        return ans


assert Solution().totalFruit([1, 2, 1]) == 3
assert Solution().totalFruit([0, 1, 2, 2]) == 3
assert Solution().totalFruit([3, 2, 7, 6]) == 2
assert Solution().totalFruit([]) == 0
assert Solution().totalFruit([1, 2, 2, 2, 3, 3, 3, 3]) == 7
assert Solution().totalFruit([1, 2, 3, 1, 2, 3]) == 2
assert Solution().totalFruit([1, 1, 2, 2, 3, 3, 3, 5, 5]) == 5
